### PR TITLE
Adds -E option to tc-print, allowing for new perf events to be used.

### PR DESCRIPTION
**Pull request by dave-estes-ampere**
The existing set of supported perf events are now designated the
predefined event types. Using the -E option replaces them with a new
set.

### CODE REVIEW

**Comment by ottoni** [comment]
@dave-estes-ampere, this change needs to be rebased.

**Comment by dave-estes-ampere** [comment]
Thanks, @ottoni. Rebased now.

File: hphp/tools/tc-print/tc-print.cpp [hunk 1]
```diff
@@ -175,7 +181,8 @@ void printValidEventTypes() {
 void parseOptions(int argc, char *argv[]) {
   int c;
   opterr = 0;
-  while ((c = getopt (argc, argv, "hc:Dd:f:g:ip:st:u:S:T:o:e:bB:v:k:a:A:n:x"))
+  char* sortByArg = NULL;
```
Comment (ottoni) [inline hphp/tools/tc-print/tc-print.cpp]: nullptr

File: hphp/tools/tc-print/tc-print.cpp [hunk 2]
```diff
@@ -334,13 +360,13 @@ void loadPerfEvents() {
   char   eventCaption[MAX_SYM_LEN];
   char   line[2*MAX_SYM_LEN];
   TCA    addr;
-  uint32_t tcSamples[NUM_EVENT_TYPES];
-  uint32_t hhvmSamples[NUM_EVENT_TYPES];
+  uint32_t tcSamples[getNumEventTypes()];
```
Comment (ottoni) [inline hphp/tools/tc-print/tc-print.cpp]: Variable-length arrays are dangerous so we don't use them. Can you rewrite to get rid of them all (eg, use std::vector)?

File: hphp/tools/tc-print/perf-events.cpp [hunk 3]
```diff
@@ -16,87 +16,84 @@
...
+  if (it == idToType.end()) {
+    idToType.insert({eventId, static_cast<PerfEventType>(numEventTypes)});
+  } else {
+    it->second = static_cast<PerfEventType>(numEventTypes);
   }
 
-  return NUM_EVENT_TYPES;
+  char caption[] = "xx";
+  sprintf(caption, "u%c",
+	  static_cast<char>(numEventTypes - getFirstEventType()) + 'A');
```
Comment (ottoni) [inline hphp/tools/tc-print/perf-events.cpp]: fix indentation (and if this is a tab, replace with whitespaces)

File: hphp/tools/tc-print/perf-events.cpp [hunk 4]
```diff
@@ -16,87 +16,84 @@
...
+  smallCaptions.push_back(std::string(caption));
+
+  numEventTypes++;
+}
+
+PerfEventType perfScriptOutputToEventType(std::string eventId) {
+  auto it = idToType.find(eventId);
+  if (it == idToType.end()) {
+    return EVENT_NULL;
+  } else {
```
Comment (ottoni) [inline hphp/tools/tc-print/perf-events.cpp]: get rid of unnecessary "else"

File: hphp/tools/tc-print/offline-code.cpp [hunk 5]
```diff
@@ -237,34 +237,18 @@ size_t OfflineCode::printBCMapping(BCMappingInfo bcMappingInfo,
...
-  const size_t NumAnnotatedEvents =
-    sizeof(AnnotatedEvents) / sizeof(AnnotatedEvents[0]);
-
-  static const char* SmallCaptions[] = {"cy", "bm", "ic", "dc", "lc", "it",
-                                        "dt"};
-
-  assert(sizeof(SmallCaptions)/sizeof(SmallCaptions[0]) == NumAnnotatedEvents);
-
-  for (size_t i = 0; i < NumAnnotatedEvents; i++) {
+				  uint32_t instrLen,
```
Comment (ottoni) [inline hphp/tools/tc-print/offline-code.cpp]: fix indentation (and get rid of potential tab)

File: hphp/tools/tc-print/offline-code.cpp [hunk 6]
```diff
@@ -237,34 +237,18 @@ size_t OfflineCode::printBCMapping(BCMappingInfo bcMappingInfo,
...
+				  const PerfEventsMap<TCA>& perfEvents) {
```
Comment (ottoni) [inline hphp/tools/tc-print/offline-code.cpp]: here again

File: hphp/tools/tc-print/tc-print.cpp [hunk 7]
```diff
@@ -252,10 +259,16 @@ void parseOptions(int argc, char *argv[]) {
...
-          exit(1);
+        // Will lookup event type after parsing all args
+        sortByArg = optarg;
+        break;
+      case 'E':
+        {
+          char* p = strtok(optarg, ",");
+          while (p) {
+            addEventType(std::string(p));
+            p = strtok(NULL, ",");
```
Comment (ottoni) [inline hphp/tools/tc-print/tc-print.cpp]: s/NULL/nullptr

**Pull request closed without merging**